### DISCUSSION TITLE
Support map_msgs/OccupancyGridUpdate

### DISCUSSION THREAD

**Discussion by tomy983** [foxglove/community#475] [Q&A]
**Description**
In 3D and 3D beta panels, while visualizing move_base costmap 2D `/move_base/local_costmap/costmap` the uptades are very slow (0.02Hz). This is probably because Studio does not also subscribe to `/move_base/local_costmap/costmap_updates`.
Confronting Studio with RVIZ, in both adding the costmap to the scene is very similar, but rviz (without notice) also subscribe to the `/move_base/local_costmap/costmap_updates` topic and most certainly uses also this topic to update the costmap on the screen as i can see using `rosnode info move_base`

- Version: 1.18.0
- OS: windows
- ROS distro (if applicable): noetic

**Steps To Reproduce**
add costmap to the 3d scene in rviz and studio. 

**Current Experience**
In Rviz the map updates are loaded into the costmap view. In studio the costmap is not updated unless a new map is received.

**Desired Experience**
The costmap in the 3D panel should take in considarition the updates coming from `/move_base/local_costmap/costmap_updates` of type map_msgs/OccupancyGridUpdate 
(same thing for global costmap..)

**Comment by defunctzombie** [comment]
@tomy983 thanks for the issue report - could you share a bag file that will help us reproduce this issue. If you'd like to share privately you can email directly or DM me in slack.

**Comment by tonynajjar** [comment]
> could you share a bag file that will help us reproduce this issue

@defunctzombie from this sentence I get the impression that costmap updates are indeed supported in Foxglove? I can't find any documentation about it. 
Essentially my costmap is huge and I would like to save bandwidth by just sending updates and having Foxglove handle the updating in the background

**Reply by karl-schulz** [reply]
I checked if the Foxglove Bridge does subscribe to the costmap updates, but it doesn't @tomy983. So this is not seem implemented at the moment.

It would be highly appreciated if this would change - using the updates instead of the full map all the time is saving a lot of bandwidth, as the costmap is basically an uncompressed bitmap.

**Reply by defunctzombie** [reply]
> @defunctzombie from this sentence I get the impression that costmap updates are indeed supported in Foxglove? I can't find any documentation about it.
> Essentially my costmap is huge and I would like to save bandwidth by just sending updates and having Foxglove handle the updating in the background

I was referring to the "the uptades are very slow (0.02Hz). " part of the original description - but I see now that this was maybe referring to the actual update rate of those messages and not the app being slow.

Indeed we have no viz support for `costmap_updates`.

Related to this costmap and costmap updates thread - is it conventional that the two appear on different topics? Rather than a single topic that has both the "full (aka keyframe)" and "update (aka delta)" data?

**Reply by tonynajjar** [reply]
> Related to this costmap and costmap updates thread - is it conventional that the two appear on different topics? Rather than a single topic that has both the "full (aka keyframe)" and "update (aka delta)" data?

I would say so yes, updates have their own [message type ](https://github.com/ros-planning/navigation_msgs/blob/rolling/map_msgs/msg/OccupancyGridUpdate.msg) and are used in Nav2

**Reply by defunctzombie** [reply]
Fascinating. Kind of the opposite of how we've supported video where its all one topic. I wonder why they did this vs a message that can tell you whether its an update or a full map. It means the tooling needs to have a way of linking the two separate topics together.

**Reply by jtbandes** [reply]
> It means the tooling needs to have a way of linking the two separate topics together.

See also: CameraInfo 😅

**Reply by acarrillo-civ** [reply]
I second that it would be great for Foxglove to match parity with `rviz2` and support subscription to the traditional nav2 `costmap_update` and `map_update` topics. I am new to the foxglove community; what is the best way to petition for this feature enhancement with the developers?

**Comment by tonynajjar** [comment]
Any update here? Is it somewhere on the roadmap?

**Comment by tonynajjar** [comment]
Is this now fully implemented in https://docs.foxglove.dev/changelog/foxglove/v2.50.0?

> 3D panel: incremental map_msgs/OccupancyGridUpdate for faster costmap-style maps

**Reply by defunctzombie** [reply]
Should be. Give it a try and let us know if it is what you expected and what else could be refined.
